### ISSUE TITLE
[Bug]: Questionable use of statics in UserService potentially causes wrong request-usersession association

### ISSUE THREAD

**Issue by phiresky** [LemmyNet/lemmy-ui#1882]
### Requirements

- [X] This is a bug report, and if not, please post to https://lemmy.ml/c/lemmy_support instead.
- [X] Please [check](https://github.com/LemmyNet/lemmy-ui/issues) to see if this issue already exists.
- [X] It's a single bug. Do not report multiple bugs in one issue.
- [X] It's a frontend issue, not a backend issue; Otherwise please create an issue on the [backend repo](https://github.com/LemmyNet/lemmy) instead.

### Summary

I heard that multiple people have allegedly experienced a situation where they load a page and are logged in as a different user - either very shortly until the page rerenders or possibly until a page refresh.

I looked at the code and it seems there might be a fairly simple explanation due to a race condition (I don't know much about Inferno, just React/SSR).

The catch all request handler:

https://github.com/LemmyNet/lemmy-ui/blob/3d58baea7cbfc4a74a7846bf94ba910cd4981d89/src/server/handlers/catch-all-handler.tsx#L25

calls initializeSite()

https://github.com/LemmyNet/lemmy-ui/blob/3d58baea7cbfc4a74a7846bf94ba910cd4981d89/src/shared/utils/app/initialize-site.ts#L5-L6

Which sets the global UserService. After that, it does an asynchronous call to `await activeRoute.fetchInitialData(initialFetchReq);` During this time, the UserService.instance.userData may be replaced by a separate incoming http request. After that, it calls `renderToString` which will create the various JSX components.

In those, the user context is used both in class componets and via the myAuth() function, e.g.:

https://github.com/LemmyNet/lemmy-ui/blob/main/src/shared/components/app/app.tsx

At which point it might be the data from a different user though.

I think the solution would be to store UserService inside the react context instead of using a singleton.

### Steps to Reproduce

I have not reproduced this issue.

### Technical Details

-

### Lemmy Instance Version

current main branch

**Comment by makotech222** [comment]
Good catch. UserService can safely be a singleton on the client, but on the SSR server it should not be.

**Comment by Die4Ever** [comment]
I would like to point out that this issue is still part of the 0.18.2 milestone, which is hidden because that milestone is closed even though this issue is not. And there are other open issues which are hidden in the same way.

https://github.com/LemmyNet/lemmy-ui/milestones?state=closed

**Comment by dessalines** [comment]
Doesn't seem to be an issue anymore.
